src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -24,10 +23,13 @@ internal class ProjectHotReloadSessionManager : IProjectHotReloadSessionManager
         private readonly Lazy<IProjectHotReloadAgent> _projectHotReloadAgent;
         private readonly Lazy<IHotReloadDiagnosticOutputService> _hotReloadDiagnosticOutputService;
 
-        private HotReloadState? _pendingSessionState = null;
-
-        private ImmutableDictionary<int, HotReloadState> _activeSessions = ImmutableDictionary<int, HotReloadState>.Empty;
+        // Protect the state from concurrent access. For example, our Process.Exited event
+        // handler may run on one thread while we're still setting up the session on
+        // another. To ensure consistent and proper behavior we need to serialize access.
+        private readonly AsyncSemaphore _semaphore = new(initialCount: 1);
```

Comment:
This is a great comment to give context. 😊

Reply:
I wrote three drafts of that comment and I find this one to be vastly better than the first and yet still not as good as I'd like it. 😭🤣